File: learning-core/executable/transaction/cpp/test_transaction.cpp
#include <cassert>
#include <unordered_map>
#include <string>

#include "transaction.cpp"

int main() {
    std::unordered_map<std::string, Account> accounts{{"a", {100}}, {"b", {10}}};
    transfer(accounts, "a", "b", 30);
    assert(accounts["a"].balance == 70);
    assert(accounts["b"].balance == 40);

    accounts["a"].balance = 20;
    accounts["b"].balance = 10;
    try {
        transfer(accounts, "a", "b", 30);
        assert(false);
    } catch (const insufficient_funds&) {
        assert(accounts["a"].balance == 20);
        assert(accounts["b"].balance == 10);
    }
}

File: learning-core/executable/transaction/python/test_transaction.py
import pytest

from transaction import Account, InsufficientFunds, transfer


def test_transfer_updates_both_sides() -> None:
    accounts = {"a": Account(100), "b": Account(10)}
    transfer(accounts, "a", "b", 30)
    assert accounts["a"].balance == 70
    assert accounts["b"].balance == 40


def test_failed_transfer_preserves_both_sides() -> None:
    accounts = {"a": Account(20), "b": Account(10)}
    with pytest.raises(InsufficientFunds):
        transfer(accounts, "a", "b", 30)
    assert accounts["a"].balance == 20
    assert accounts["b"].balance == 10
